File: day-1/Simulator.java
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class Simulator {
    private static final int MAX_PITY = 10;
    private static final int MAX_GUARANTEE = 120;

    private int totalPulls;
    private int pity;
    private int guarantee;
    private int legendaryCount;
    private int featuredCount;
    private final Random random = new Random();

    public PullResult pull() {
        boolean featuredGuaranteed = guarantee + 1 >= MAX_GUARANTEE;
        boolean pityGuaranteed = pity + 1 >= MAX_PITY;

        PullResult result = new PullResult(totalPulls + 1, "Rare", false, false);

        if (featuredGuaranteed) {
            result = new PullResult(totalPulls + 1, "Featured", false, true);
        } else if (pityGuaranteed) {
            result = new PullResult(totalPulls + 1, "Legendary", true, false);
        } else {
            int roll = random.nextInt(100) + 1;

            if (roll <= 70) {
                result = new PullResult(totalPulls + 1, "Rare", false, false);
            } else if (roll <= 95) {
                result = new PullResult(totalPulls + 1, "Epic", false, false);
            } else {
                result = new PullResult(totalPulls + 1, "Legendary", false, false);
            }
        }

        applyResult(result);
        return result;
    }

    public List<PullResult> pullMany(int amount) {
        List<PullResult> results = new ArrayList<>();

        for (int index = 0; index < amount; index++) {
            results.add(pull());
        }

        return results;
    }

    public void printStats() {
        System.out.println();
        System.out.println("Total pulls: " + totalPulls);
        System.out.println("Pity: " + pity + " / " + MAX_PITY);
        System.out.println("Featured guarantee: " + guarantee + " / " + MAX_GUARANTEE);
        System.out.println("Legendary count: " + legendaryCount);
        System.out.println("Featured count: " + featuredCount);
    }

    private void applyResult(PullResult result) {
        totalPulls++;

        boolean isLegendary = result.rarity().equals("Legendary");
        boolean isFeatured = result.rarity().equals("Featured");

        pity = isLegendary || isFeatured ? 0 : pity + 1;
        guarantee = isFeatured ? 0 : guarantee + 1;
        legendaryCount += isLegendary ? 1 : 0;
        featuredCount += isFeatured ? 1 : 0;
    }

    public static void main(String[] args) {
        Simulator simulator = new Simulator();
        List<PullResult> results = simulator.pullMany(10);

        for (PullResult result : results) {
            System.out.print("#" + result.number() + " " + result.rarity());

            if (result.pityTriggered()) {
                System.out.print(" (pity)");
            }

            if (result.featuredTriggered()) {
                System.out.print(" (120 guarantee)");
            }

            System.out.println();
        }

        simulator.printStats();
    }

    public record PullResult(
        int number,
        String rarity,
        boolean pityTriggered,
        boolean featuredTriggered
    ) {
    }
}

File: day-1/simulator.cpp
#include <iostream>
#include <random>
#include <string>
#include <vector>

constexpr int MaxPity = 10;
constexpr int MaxGuarantee = 120;

struct PullResult {
  int number;
  std::string rarity;
  bool pityTriggered;
  bool featuredTriggered;
};

class PityPullSimulator {
 public:
  PityPullSimulator() : rng(std::random_device{}()), roll(1, 100) {}

  PullResult pull() {
    const bool featuredGuaranteed = guarantee + 1 >= MaxGuarantee;
    const bool pityGuaranteed = pity + 1 >= MaxPity;

    PullResult result{
      totalPulls + 1,
      "Rare",
      false,
      false
    };

    if (featuredGuaranteed) {
      result.rarity = "Featured";
      result.featuredTriggered = true;
    } else if (pityGuaranteed) {
      result.rarity = "Legendary";
      result.pityTriggered = true;
    } else {
      const int value = roll(rng);
      if (value <= 70) {
        result.rarity = "Rare";
      } else if (value <= 95) {
        result.rarity = "Epic";
      } else {
        result.rarity = "Legendary";
      }
    }

    applyResult(result);
    return result;
  }

  std::vector<PullResult> pullMany(int amount) {
    std::vector<PullResult> results;
    results.reserve(amount);

    for (int index = 0; index < amount; index += 1) {
      results.push_back(pull());
    }

    return results;
  }

  void printStats() const {
    std::cout << "\nTotal pulls: " << totalPulls
              << "\nPity: " << pity << " / " << MaxPity
              << "\nFeatured guarantee: " << guarantee << " / " << MaxGuarantee
              << "\nLegendary count: " << legendaryCount
              << "\nFeatured count: " << featuredCount << "\n";
  }

 private:
  int totalPulls = 0;
  int pity = 0;
  int guarantee = 0;
  int legendaryCount = 0;
  int featuredCount = 0;
  std::mt19937 rng;
  std::uniform_int_distribution<int> roll;

  void applyResult(const PullResult& result) {
    totalPulls += 1;

    const bool isLegendary = result.rarity == "Legendary";
    const bool isFeatured = result.rarity == "Featured";

    pity = isLegendary || isFeatured ? 0 : pity + 1;
    guarantee = isFeatured ? 0 : guarantee + 1;
    legendaryCount += isLegendary ? 1 : 0;
    featuredCount += isFeatured ? 1 : 0;
  }
};

int main() {
  PityPullSimulator simulator;
  const auto results = simulator.pullMany(10);

  for (const PullResult& result : results) {
    std::cout << "#" << result.number << " " << result.rarity;

    if (result.pityTriggered) {
      std::cout << " (pity)";
    }

    if (result.featuredTriggered) {
      std::cout << " (120 guarantee)";
    }

    std::cout << "\n";
  }

  simulator.printStats();
  return 0;
}
